Rename vmm_vendor -> vmm_vendor_display

imtayadeway opened the pull request:
Provides a better fix for https://github.com/ManageIQ/manageiq/commit/76e9afe3e0bc1f168067e11afba9d35ae61c1af8, and fixes a few issues in the code that was touched as part of the rename. In particular:

* Fixes a code path that was unreachable (checked for `vmm_vendor == "vmware"` - old `vmm_vendor` or the new `vmm_vendor_display` would be "VmWare")
* Removes a MiqSearch that was no longer needed.

imtayadeway commented on app/controllers/application_controller.rb:
```diff
@@ -1499,11 +1499,11 @@ def set_config(db_record) # pass in the db record, either @host or @vm
 
       @vmminfo.push(:vmminfo     => "Vendor",
-                    :description => db_record.vmm_vendor) unless db_record.vmm_vendor.nil?
+                    :description => db_record.vmm_vendor_display) unless db_record.vmm_vendor_display.nil? # This would never have been nil
```
If the record is valid, then this can't be `nil`. Am I safe just to remove this condition?

Fryguy replied:
Yeah, the respond_to? up above handles it.

imtayadeway commented on app/models/host.rb:
```diff
@@ -908,7 +908,7 @@ def reset_discoverable_fields
     # Determine os
     os_type = nil
-    if vmm_vendor == "vmware"
+    if vmm_vendor_display == "vmware" # ??? This will never be true
```
I don't think this can ever be true. Is this path still needed?

Fryguy replied:
detect_discovered_os, I believe, is for bare-metal Host provisioning.  @gmcculloug @bdunne Can you comment here?

Fryguy replied:
should this just be a `is_vmware?` check like the others?

imtayadeway replied:
@Fryguy :+1: I will use that method if that's right....just concerned that AFAICS that path wasn't executing before, so wondered if was still needed.

bdunne replied:
@Fryguy yes, I believe it is used for bare-metal and can be `is_vmware?` @gmcculloug please confirm

imtayadeway commented on db/fixtures/miq_searches.yml:
```diff
@@ -1,103 +1,103 @@
 ---
-- attributes: 
```
My editor killed the trailing whitespace in this and other yml files - I'll put these back if it's considered too much noise.

Fryguy replied:
I think it's fine...if it's not a lot of effort to put back, then yeah, maybe undo it and have a follow up PR, but it's up to you.

imtayadeway commented:
Here's a grep of everything that wasn't converted `vmm_vendor` -> `vmm_vendor_display`:

```
tim@tim-thinkpad:~/src/manageiq$ ag vmm_vendor | grep -v vmm_vendor_display
db/migrate/20130923182042_collapsed_initial_migration.rb:727:      t.string   "vmm_vendor"
db/fixtures/miq_searches.yml:163:          field: Host-vmm_vendor
db/fixtures/miq_searches.yml:175:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:192:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:209:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:226:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:243:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:260:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:276:          field: Host-vmm_vendor
config/locales/en.yml:612:      vmm_vendor: VMM Vendor
spec/models/rbac_spec.rb:200:      it "can filter results by vmm_vendor" do
spec/models/rbac_spec.rb:201:        host = FactoryGirl.create(:host, :vmm_vendor => "vmware")
spec/models/rbac_spec.rb:202:        expression = MiqExpression.new("=" => {"field" => "Host-vmm_vendor", "value" => "vmware"})
app/models/host.rb:41:  validates_inclusion_of    :vmm_vendor, :in => VENDOR_TYPES.keys
app/models/host.rb:643:    VENDOR_TYPES[vmm_vendor]
app/models/host.rb:650:    self[:vmm_vendor] = v
app/models/host.rb:894:      next if %w(id guid ipmi_address mac_address name created_on updated_on vmm_vendor).include?(key)
```

I left the search stuff as `vmm_vendor` specifically so Rbac could take advantage of searching for hosts in SQL, making this a better fix for https://bugzilla.redhat.com/show_bug.cgi?id=1291854 (see https://github.com/ManageIQ/manageiq/pull/6207).

imtayadeway commented:
/cc @matthewd

Fryguy commented on app/models/host.rb:
```diff
@@ -610,7 +610,7 @@ def is_scannable_now_error_message
 
   def is_vmware?
-    vmm_vendor.to_s.strip.downcase == 'vmware'
+    vmm_vendor_display.to_s.strip.downcase == 'vmware'
```
I feel like this will make a lot easier to get rid of these terrible conditionals. :+1:

Fryguy commented on app/models/manageiq/providers/vmware/infra_manager/refresh_parser.rb:
```diff
@@ -188,34 +188,34 @@ def self.host_inv_to_hashes(inv, ems_inv, storage_uids, cluster_uids)
+            :hostname           => hostname,
+            :ipaddress          => ipaddress,
+            :uid_ems            => uid_ems,
+            :vmm_vendor_display => vendor,
```
I feel like the parsers should still set the column directly, so this should still set `vmm_vendor`

Fryguy commented on app/models/manageiq/providers/redhat/infra_manager/refresh_parser.rb:
```diff
@@ -110,23 +110,23 @@ def self.host_inv_to_hashes(inv, ems_inv, cluster_uids, _storage_uids)
+        :hostname           => hostname,
+        :ipaddress          => ipaddress,
+        :uid_ems            => host_inv[:id],
+        :vmm_vendor_display => 'RedHat',
```
Same

imtayadeway commented on db/fixtures/miq_searches.yml:
```diff
@@ -154,47 +154,47 @@
+        "=":
           field: Host-vmm_vendor
-          value: EC2
+          value: EC2 # Is this a valid vendor?
```
Is this search still needed? Looks like this can no longer be a valid `vmm_vendor` value (regardless of case), unless I'm mistaken.

Fryguy replied:
I think Amazon stuff doesn't exist anymore, so this can probably be removed.  @blomquisg or @gmcculloug  keep me honest here.

Fryguy commented on spec/factories/host.rb:
```diff
@@ -2,7 +2,7 @@
     sequence(:name)     { |n| "host_#{seq_padded_for_sorting(n)}" }
     sequence(:hostname) { |n| "host_#{seq_padded_for_sorting(n)}" }
-    vmm_vendor          "vmware"
+    vmm_vendor_display  "VMware"
```
I think the setter for vmm_vendor_display is weird.  Ideally, the code should work with the raw column, and vmm_vendor_display should only be used for display / UI purposes.  Having "display" knowledge spread around (like in factories and in the refresh parser, feels wrong).

imtayadeway replied:
Yeah, me too. I wasn't sure what I might break, so I went conservatively with updating most things to use the method that had the same behavior as before. I can remove the setter (and update accordingly) if you think that wouldn't cause any headaches.

Fryguy replied:
Yeah, remove the setter...I think it makes it a lot clearer which code needs which responsibility, and which code is doing the Wrong Thing :tm:

imtayadeway replied:
Or I could leave the setter as `#vmm_vendor=`?

Fryguy replied:
I think the callers for setting vmm_vendor should be changed to call it with the raw value...

For example,
before: `blah.vmm_vendor = 'Microsoft'`
after: `blah.vmm_vendor = 'microsoft'`
as opposed to `blah.vmm_vendor_display = 'Microsoft'`

Then you don't need the old `#vmm_vendor=` logic at all

imtayadeway replied:
:thumbsup: thanks!

Fryguy commented on spec/models/manageiq/providers/microsoft/infra_manager/refresher_spec.rb:
```diff
@@ -115,15 +115,15 @@ def assert_specific_cluster
+      :name               => "hyperv-h01.manageiq.com",
+      :hostname           => "hyperv-h01.manageiq.com",
+      :ipaddress          => "192.168.252.99",
+      :vmm_vendor_display => "Microsoft",
```
Refresher specs shouldn't care about `vmm_vendor_display` either, since they are not setting them.   They should only care about `vmm_vendor`

Fryguy commented on spec/models/manageiq/providers/openstack/infra_manager/refresher_rhos_juno_spec.rb:
```diff
@@ -95,13 +95,13 @@ def assert_specific_host
-      :connection_state => "connected",
-      :service_tag      => nil,
+      :ipmi_address       => nil,
+      :vmm_vendor_display => "RedHat",
```
Same

Fryguy commented on spec/models/miq_report_spec.rb:
```diff
@@ -89,8 +89,8 @@ def invoke_ae
-    result = [{"id" => 5, "vmm_vendor" => "VMware", "vmm_product" => "ESXi", "ipaddress" => "192.168.252.13", "vmm_buildnumber" => "260247", "vmm_version" => "4.1.0", "name" => "VI4ESXM1.manageiq.com"}, {"id" => 3, "vmm_vendor" => "VMware", "vmm_product" => "ESXi", "ipaddress" => "192.168.252.9", "vmm_buildnumber" => "348481", "vmm_version" => "4.1.0", "name" => "vi4esxm2.manageiq.com"}, {"id" => ...
-    column_names = ["name", "ipaddress", "vmm_vendor", "vmm_product", "vmm_version", "vmm_buildnumber", "id"]
+    result = [{"id" => 5, "vmm_vendor_display" => "VMware", "vmm_product" => "ESXi", "ipaddress" => "192.168.252.13", "vmm_buildnumber" => "260247", "vmm_version" => "4.1.0", "name" => "VI4ESXM1.manageiq.com"}, {"id" => 3, "vmm_vendor" => "VMware", "vmm_product" => "ESXi", "ipaddress" => "192.168.252.9", "vmm_buildnumber" => "348481", "vmm_version" => "4.1.0", "name" => "vi4esxm2.manageiq.com"},  ...
+    column_names = ["name", "ipaddress", "vmm_vendor_display", "vmm_product", "vmm_version", "vmm_buildnumber", "id"]
```
This should probably test both vmm_vendor and vmm_vendor_display.

matthewd commented:
FYI, this is now blocking https://github.com/ManageIQ/manageiq/pull/5982, unless I revert 76e9afe3e0bc1f168067e11afba9d35ae61c1af8.

imtayadeway commented:
@gmcculloug @bdunne @blomquisg could I have some feedback on @Fryguy's comments?
@matthewd alternatively I could just proceed with rebasing this and following up with another PR if this is blocking. I *think* it should be GTG otherwise, the other proposed changes aren't critical

imtayadeway commented:
Just updated and fixed the above issues in https://github.com/ManageIQ/manageiq/commit/b135639adff0e076ecc1a4e6214ccc09dab59d6e and https://github.com/imtayadeway/manageiq/commit/eaafc46a9e198545971aea646bec8cc2ecb1aec2

Here's the new list of things that are still `vmm_vendor`:

```
tim@tim-thinkpad:~/src/manageiq$ ag vmm_vendor | grep -v vmm_vendor_display
db/migrate/20130923182042_collapsed_initial_migration.rb:727:      t.string   "vmm_vendor"
db/fixtures/miq_searches.yml:164:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:181:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:198:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:215:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:232:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:249:            field: Host-vmm_vendor
db/fixtures/miq_searches.yml:265:          field: Host-vmm_vendor
config/locales/en.yml:597:      vmm_vendor: VMM Vendor
spec/models/rbac_spec.rb:200:      it "can filter results by vmm_vendor" do
spec/models/rbac_spec.rb:201:        host = FactoryGirl.create(:host, :vmm_vendor => "vmware")
spec/models/rbac_spec.rb:202:        expression = MiqExpression.new("=" => {"field" => "Host-vmm_vendor", "value" => "vmware"})
spec/models/vim_performance_tag_spec.rb:22:      @host = FactoryGirl.create(:host, :vms => @vms, :vmm_vendor => 'vmware', :vmm_product => "ESX", :type => "ManageIQ::Providers::Vmware::InfraManager::HostEsx")
spec/models/job_proxy_dispatcher_vm_storage2hosts_spec.rb:77:            @host.vmm_vendor = "microsoft"
spec/models/manageiq/providers/vmware/infra_manager/refresher_spec.rb:179:      :vmm_vendor         => "vmware",
spec/models/manageiq/providers/redhat/infra_manager/refresher_3_1_spec.rb:156:      :vmm_vendor       => "redhat",
spec/models/manageiq/providers/redhat/infra_manager/refresher_3_0_spec.rb:140:      :vmm_vendor       => "redhat",
spec/models/manageiq/providers/openstack/infra_manager/refresher_rhos_juno_spec.rb:99:      :vmm_vendor       => "redhat",
spec/models/manageiq/providers/microsoft/infra_manager/refresher_spec.rb:122:      :vmm_vendor       => "microsoft",
spec/models/miq_host_provision_spec.rb:17:    mhp.host = FactoryGirl.create(:host, :vmm_vendor => 'unknown')
spec/models/miq_host_provision_spec.rb:20:    mhp.host.vmm_vendor = 'vmware'
spec/lib/ems_event_helper_spec.rb:20:                                      :vmm_vendor            => 'vmware',
spec/factories/host.rb:5:    vmm_vendor          "vmware"
spec/factories/host.rb:50:    vmm_vendor "redhat"
spec/factories/host.rb:54:    vmm_vendor  "unknown"
spec/factories/host.rb:60:    vmm_vendor  "microsoft"
app/controllers/host_controller.rb:251:      @host.vmm_vendor = "unknown"
app/models/host.rb:41:  validates_inclusion_of    :vmm_vendor, :in => VENDOR_TYPES.keys
app/models/host.rb:613:    vmm_vendor == 'vmware'
app/models/host.rb:643:    VENDOR_TYPES[vmm_vendor]
app/models/host.rb:887:      next if %w(id guid ipmi_address mac_address name created_on updated_on vmm_vendor).include?(key)
app/models/host.rb:894:    self.vmm_vendor = 'unknown'
app/models/host.rb:924:      self.vmm_vendor  = "microsoft"
app/models/host.rb:929:      self.vmm_vendor  = "vmware"
app/models/host.rb:949:      self.vmm_vendor   = "unknown"
app/models/host.rb:955:      self.vmm_vendor = ost.hypervisor.join(", ")
app/models/manageiq/providers/vmware/infra_manager/refresh_parser.rb:198:            :vmm_vendor       => vendor,
app/models/manageiq/providers/redhat/infra_manager/refresh_parser.rb:120:        :vmm_vendor       => 'redhat',
app/models/manageiq/providers/openstack/infra_manager/refresh_parser.rb:196:          :vmm_vendor           => 'redhat',
app/models/manageiq/providers/microsoft/infra_manager/refresh_parser.rb:118:        :vmm_vendor       => 'microsoft',
```

I *think* that's right. Only thing I'm not sure about is:

```
config/locales/en.yml:597:      vmm_vendor: VMM Vendor
```

Otherwise this should be GTG.

/cc @matthewd

imtayadeway commented:
@miq-bot rm-label wip

chessbyte commented:
@imtayadeway Why are we renaming this column/method?  Performance?

imtayadeway commented:
@chessbyte it's to provide a better fix for https://github.com/ManageIQ/manageiq/commit/76e9afe3e0bc1f168067e11afba9d35ae61c1af8 - made more urgent because @matthewd will have to revert that fix to continue with Rails 5 work unless this gets merged first.

Fryguy commented:
@imtayadeway This looks great.  Can you fix the alignment changes you made in spec/models/manageiq/providers/vmware/infra_manager/refresher_spec.rb as per the bot above?  None of those alignments need to have changed.

Otherwise this is ready to merge, from my perspective.

imtayadeway commented:
@Fryguy Just fixed the offending commit and repushed. Thanks!
